Refactored properties raise key error if index is set in dataframe

## 🐛 Bug

<!-- A clear and concise description of what the bug is. -->

## To Reproduce

Steps to reproduce the behavior:

This test passes on 0.4.1.2 but fails on the current main (this is short reproduction of code from PartSeg):

```python
def test_get_status():
    data = np.zeros((10, 10), dtype=np.uint8)
    data[2:5, 2:-2] = 1
    data[5:-2, 2:-2] = 2
    layer = Labels(data)
    layer.properties = pd.DataFrame({"text1": [1, 3], "text2": [7, -2], "index" : [1, 2]}, index=[1, 2])
    assert layer.get_status((0, 0)) == "Labels [0 0]: 0; [No Properties]"
    assert layer.get_status((3, 3)) == "Labels [3 3]: 1; text1: 1, text2: 7"
    assert layer.get_status((6, 6)) == "Labels [6 6]: 2; text1: 3, text2: -2"
```

<!-- If you have a code sample, error messages, stack traces, please provide it here as well -->

```
napari/layers/labels/_tests/test_labels.py::test_get_status FAILED       [100%]
napari/layers/labels/_tests/test_labels.py:1347 (test_get_status)
def test_get_status():
        data = np.zeros((10, 10), dtype=np.uint8)
        data[2:5, 2:-2] = 1
        data[5:-2, 2:-2] = 2
        layer = Labels(data)
        layer.properties = pd.DataFrame({"text1": [1, 3], "text2": [7, -2], "index" : [1, 2]}, index=[1, 2])
        assert layer.get_status((0, 0)) == "Labels [0 0]: 0; [No Properties]"
>       assert layer.get_status((3, 3)) == "Labels [3 3]: 1; text1: 1, text2: 7"

napari/layers/labels/_tests/test_labels.py:1355: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
napari/layers/labels/labels.py:1345: in get_status
    properties = self._get_properties(
napari/layers/labels/labels.py:1415: in _get_properties
    return [
napari/layers/labels/labels.py:1420: in <listcomp>
    and v[idx] is not None
../../.pyenv/versions/napari/lib/python3.8/site-packages/pandas/core/series.py:871: in __getitem__
    result = self.index.get_value(self, key)
../../.pyenv/versions/napari/lib/python3.8/site-packages/pandas/core/indexes/base.py:4405: in get_value
    return self._engine.get_value(s, k, tz=getattr(series.dtype, "tz", None))
pandas/_libs/index.pyx:80: in pandas._libs.index.IndexEngine.get_value
    ???
pandas/_libs/index.pyx:90: in pandas._libs.index.IndexEngine.get_value
    ???
pandas/_libs/index.pyx:138: in pandas._libs.index.IndexEngine.get_loc
    ???
pandas/_libs/hashtable_class_helper.pxi:998: in pandas._libs.hashtable.Int64HashTable.get_item
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   KeyError: 0

pandas/_libs/hashtable_class_helper.pxi:1005: KeyError
```

## Expected behavior
The text, may change, but the code should work, or at least not end with an error. 

<!-- A clear and concise description of what you expected to happen. -->

## Environment

last main

## Additional context

<!-- Add any other context about the problem here. -->

--- a/napari/layers/labels/_tests/test_labels.py
+++ b/napari/layers/labels/_tests/test_labels.py
@@ -1344,3 +1344,18 @@ def test_negative_label_doesnt_flicker():
     # the label colors. Now -1 is seen so it is taken into account in the
     # indexing calculation, and changes color
     assert tuple(layer.get_color(-1)) == minus_one_color_original
+
+
+def test_get_status_with_custom_index():
+    """See https://github.com/napari/napari/issues/3811"""
+    data = np.zeros((10, 10), dtype=np.uint8)
+    data[2:5, 2:-2] = 1
+    data[5:-2, 2:-2] = 2
+    layer = Labels(data)
+    df = pd.DataFrame(
+        {'text1': [1, 3], 'text2': [7, -2], 'index': [1, 2]}, index=[1, 2]
+    )
+    layer.properties = df
+    assert layer.get_status((0, 0)) == 'Labels [0 0]: 0; [No Properties]'
+    assert layer.get_status((3, 3)) == 'Labels [3 3]: 1; text1: 1, text2: 7'
+    assert layer.get_status((6, 6)) == 'Labels [6 6]: 2; text1: 3, text2: -2'

--- a/napari/layers/utils/_tests/test_layer_utils.py
+++ b/napari/layers/utils/_tests/test_layer_utils.py
@@ -432,3 +432,17 @@ def test_remove_features():
         new_features['confidence'],
         [0.2, 1],
     )
+
+
+def test_validate_features_with_custom_index_without_num_data():
+    input = pd.DataFrame({'a': [1, 3], 'b': [7.5, -2.1]}, index=[1, 2])
+    actual = _validate_features(input)
+    expected = input.reset_index(drop=True)
+    pd.testing.assert_frame_equal(actual, expected)
+
+
+def test_validate_features_with_custom_index_and_num_data():
+    input = pd.DataFrame({'a': [1, 3], 'b': [7.5, -2.1]}, index=[1, 2])
+    actual = _validate_features(input, num_data=2)
+    expected = input.reset_index(drop=True)
+    pd.testing.assert_frame_equal(actual, expected)

--- a/napari/layers/utils/layer_utils.py
+++ b/napari/layers/utils/layer_utils.py
@@ -695,8 +695,9 @@ def _validate_features(
     ----------
     features : Union[Dict[str, np.ndarray], pd.DataFrame]
         The features table input, which will be passed to the pandas
-        DataFrame initializer. If this has a non-default index, it
-        will be ignored.
+        DataFrame initializer. If this is a pandas DataFrame with a
+        non-default index, that index (except its length) will be
+        ignored.
     num_data : Optional[int]
         The number of the elements in the layer calling this, such as
         the number of points, which is used to check that the features
@@ -706,9 +707,8 @@ def _validate_features(
     Returns
     -------
     pd.DataFrame
-        The pandas DataFrame created from the input features table.
-        In general, this is will copy the features data, though that
-        behavior should not be relied on.
+        The pandas DataFrame created from the input features table. This
+        may or may not store a copy of the input data.
 
     Raises
     ------

--- a/napari/layers/utils/layer_utils.py
+++ b/napari/layers/utils/layer_utils.py
@@ -695,7 +695,8 @@ def _validate_features(
     ----------
     features : Union[Dict[str, np.ndarray], pd.DataFrame]
         The features table input, which will be passed to the pandas
-        DataFrame initializer.
+        DataFrame initializer. If this has a non-default index, it
+        will be ignored.
     num_data : Optional[int]
         The number of the elements in the layer calling this, such as
         the number of points, which is used to check that the features
@@ -706,8 +707,8 @@ def _validate_features(
     -------
     pd.DataFrame
         The pandas DataFrame created from the input features table.
-        If the input features are already a DataFrame, the data will not
-        be copied, otherwise they will.
+        In general, this is will copy the features data, though that
+        behavior should not be relied on.
 
     Raises
     ------
@@ -715,6 +716,8 @@ def _validate_features(
         If the input feature columns are not all the same length, or if
         that length is not equal to the given num_data.
     """
+    if isinstance(features, pd.DataFrame):
+        features = features.reset_index(drop=True)
     index = None if num_data is None else range(num_data)
     return pd.DataFrame(data=features, index=index)
 

